iter/iter_test.go

```diff
@@ -2,12 +2,82 @@ package iter
 
 import (
 	"strconv"
+	"sync/atomic"
 	"testing"
+	"time"
 
 	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
+func TestIterator(t *testing.T) {
+	t.Parallel()
+
+	t.Run("safe for reuse", func(t *testing.T) {
+		t.Parallel()
+
+		iterator := Iterator[int]{MaxGoroutines: 999}
+
+		// iter.Concurrency > numInput case that updates iter.Concurrency
+		iterator.ForEachIdx([]int{1, 2, 3}, func(i int, t *int) {})
+
+		assert.Equal(t, iterator.MaxGoroutines, 999)
+	})
+
+	t.Run("allows more than defaultMaxGoroutines() concurrent tasks", func(t *testing.T) {
+		t.Parallel()
+
+		wantConcurrency := 2 * defaultMaxGoroutines()
+
+		testDone, forEachDone := make(chan struct{}), make(chan struct{})
+
+		var concurrentTasks atomic.Int64
+		go func() {
+			// Run in a goroutine because we don't allow the callbacks
+			// to return until the conclusion of the test, so ForEach
+			// will block.
+			tasks := make([]int, wantConcurrency)
+			iterator := Iterator[int]{MaxGoroutines: wantConcurrency}
+
+			iterator.ForEach(tasks, func(t *int) {
+				concurrentTasks.Add(1)
+				// Block until conclusion of test. This ensures that
+				// all jobs must be submitted, despite the input
+				// being larger than runtime.GOMAXPROCS(0)
+				<-testDone
+			})
+
+			// Signal that iterator.ForEach has exited.
+			forEachDone <- struct{}{}
+		}()
+
+		// Wait up to ~3 seconds for all jobs to be submitted. Realistically,
+		// this will take less than a second.
+		var ms int
+		for ms < 3000 && concurrentTasks.Load() < int64(wantConcurrency) {
+			select {
+			case <-forEachDone:
+				t.Error("iterator.ForEach exited before tasks were all submitted")
+			default:
+				ms += 1
+				time.Sleep(time.Millisecond)
+			}
+		}
+		assert.Equal(t, concurrentTasks.Load(), int64(wantConcurrency))
+
+		// Allow all tasks to return, and make sure iterator.ForEach also
+		// returns.
+		close(testDone)
+		select {
+		case <-forEachDone:
+		case <-time.After(1 * time.Second):
+			t.Error("iterator.ForEach did not exit within 1 second")
+		}
+		close(forEachDone)
+	})
```

Comment:
I think we can significantly simplify this test. We can depend on test timeouts to detect deadlocks (rather than enforcing timeouts manually). This also has the added benefit that the failure will dump a stacktrace. WDYT?

```suggestion
	t.Run("allows more than defaultMaxGoroutines() concurrent tasks", func(t *testing.T) {
		t.Parallel()

		wantConcurrency := 2 * defaultMaxGoroutines()

		maxConcurrencyHit := make(chan struct{})

		tasks := make([]int, wantConcurrency)
		iterator := Iterator[int]{MaxGoroutines: wantConcurrency}

		var concurrentTasks atomic.Int64
		iterator.ForEach(tasks, func(t *int) {
			n := concurrentTasks.Add(1)
			defer concurrentTasks.Add(-1)

			if int(n) == wantConcurrency {
				// All our tasks are running concurrently.
				// Signal to the rest of the tasks to stop.
				close(maxConcurrencyHit)
			} else {
				// Wait until we hit max concurrency before exiting
				<-maxConcurrencyHit
			}
		})
	})
```

Reply:
I originally thought about having failures be enforced by test timeouts, but I figured the test should assert that this happens ~instantly - I'm not sure what's more idiomatic 🤔

That said, what you've written here is much cleaner, thanks!

Reply:
> assert that this happens ~instantly

If it doesn't happen ~instantly, it's probably a deadlock, which test timeouts will catch. I don't think we'll ever be in a situation where our timeout is accurate enough to consistently tell the difference between "not running in parallel" and "running in parallel" so I think it's probably best to just avoid any chance of flakiness.